giraffe/src/transforms/band.ts

```diff
@@ -0,0 +1,210 @@
+      } else if (line[RESULT] === MIN) {
+        bandIndices.minIndices.push(index)
+      } else {
+        bandIndices.rowIndices.push(index)
+      }
+    })
+  }
+
+  return bandIndices
+}
+
+export const getBandLineIndices = (fillColumnMap: ColumnGroupMap): number[] => {
+  const bandLineIndices = Array.isArray(fillColumnMap.mappings)
+    ? fillColumnMap.mappings.reduce((indices, line, index) => {
+        if (line[RESULT] && line[RESULT] !== MAX && line[RESULT] !== MIN) {
+          indices.push(index)
+        }
+        return indices
+      }, [])
+    : []
+
+  return bandLineIndices as number[]
+}
+
+export const getBandName = (fillColumnMap: ColumnGroupMap): string => {
+  return Array.isArray(fillColumnMap.mappings)
+    ? fillColumnMap.mappings.reduce((name, line) => {
+        if (!name && line[RESULT] !== MAX && line[RESULT] !== MIN) {
+          name = line[RESULT]
+        }
+        return name
+      }, '')
+    : ''
+}
+
+export const mapCumulativeValuesToTimeRange = (
+  timesCol: NumericColumnData,
+  valuesCol: NumericColumnData,
+  fillCol: NumericColumnData
+): CumulativeValuesByTime => {
+  const cumulativeValues: CumulativeValuesByTime = {}
+
+  timesCol.forEach((time, index) => {
+    if (!cumulativeValues[time]) {
+      cumulativeValues[time] = {}
+    }
+    const maxGroup = fillCol[index]
+
+    if (!cumulativeValues[time][maxGroup - 1]) {
+      for (let i = 0; i < maxGroup; i += 1) {
+        if (!cumulativeValues[time][i]) {
+          cumulativeValues[time][i] = 0
+        }
+      }
+    }
+    cumulativeValues[time][maxGroup] =
+      maxGroup - 1 >= 0
+        ? cumulativeValues[time][maxGroup - 1] + valuesCol[index]
+        : valuesCol[index]
+  })
+  return cumulativeValues
+}
+
+export const bandTransform = (
+  inputTable: Table,
+  xColumnKey: string,
+  yColumnKey: string,
+  fillColKeys: string[],
+  colors: string[]
+): BandLayerSpec => {
+  const [fillColumn, fillColumnMap] = createGroupIDColumn(
+    inputTable,
+    fillColKeys
+  )
+
+  const table = inputTable.addColumn(FILL, 'number', fillColumn)
+  const xCol = table.getColumn(xColumnKey, 'number')
+  const yCol = table.getColumn(yColumnKey, 'number')
+  const fillScale = getNominalColorScale(fillColumnMap, colors)
+  const bandFillColors = []
+  const lineData: LineData = {}
+
+  let xMin = Infinity
+  let xMax = -Infinity
+  let yMin = Infinity
+  let yMax = -Infinity
+
+  for (let i = 0; i < table.length; i++) {
+    const groupID = fillColumn[i]
+    const x = xCol[i]
+    const y = yCol[i]
+
+    if (!lineData[groupID]) {
+      lineData[groupID] = {xs: [], ys: [], fill: fillScale(groupID)}
+      bandFillColors.push(fillScale(groupID))
+    }
+
+    lineData[groupID].xs.push(x)
+    lineData[groupID].ys.push(y)
+
+    if (x < xMin) {
+      xMin = x
+    }
+
+    if (x > xMax) {
+      xMax = x
+    }
+
+    if (y < yMin) {
+      yMin = y
+    }
+
+    if (y > yMax) {
+      yMax = y
+    }
```

Comment:
alternatively, this can be expressed in 4 lines:

```ts
xMin = Math.min(x, xMin)
xMax = Math.max(x, xMax)
yMin = Math.min(y, yMin)
yMax = Math.max(y, yMax)
```

Reply:
Yes, I like this. 👍